server/src/main/java/org/elasticsearch/gateway/IncrementalClusterStateWriter.java

```diff
@@ -104,19 +107,118 @@ void setIncrementalWrite(boolean incrementalWrite) {
+        final long startTimeMillis = relativeTimeMillisSupplier.getAsLong();
+
+        final AtomicClusterStateWriter writer;
+        synchronized (this) {
+            indicesToRetainOnDataNode = IncrementalClusterStateWriter.getRelevantIndicesOnDataOnlyNode(clusterState);
+            writer = new AtomicClusterStateWriter(metaStateService, previousManifest);
+            final Map<Index, Long> indexGenerations = new HashMap<>(previousManifest.getIndexGenerations());
+            if (indexGenerations.entrySet().removeIf(entry -> indicesToRetainOnDataNode.contains(entry.getKey()) == false)) {
+                Manifest manifest = new Manifest(previousManifest.getCurrentTerm(), previousManifest.getClusterStateVersion(),
+                    previousManifest.getGlobalGeneration(), indexGenerations);
+                assert manifest.getIndexGenerations().keySet().stream().allMatch(indicesToRetainOnDataNode::contains);
+                writeManifest(writer, manifest);
+                previousManifest = manifest;
+            }
+        }
+
+        logSlowWrites(startTimeMillis, writer);
+    }
+
+    /**
+     * Updates the manifest and metadata on disk for data-only nodes. As this method will run concurrently to clean-up actions on the
+     * node, we have to make sure that we always check whether we should still work on a given index, which is updated
+     * by {@link #retainIndicesOnDataOnlyNode}.
+     *
+     * @param newState new {@link ClusterState}
+     * @throws WriteStateException if exception occurs. See also {@link WriteStateException#isDirty()}.
+     */
+    void updateClusterStateForDataOnlyNode(ClusterState newState)
+        throws WriteStateException {
+        assert newState.nodes().getLocalNode().isMasterNode() == false && newState.nodes().getLocalNode().isDataNode();
+        final long startTimeMillis = relativeTimeMillisSupplier.getAsLong();
+        final AtomicClusterStateWriter writer;
+        final long globalStateGeneration;
+        final Iterable<IndexMetaDataAction> actions;
+        final Manifest startManifest;
+        synchronized (this) {
+            startManifest = previousManifest;
+            writer = new AtomicClusterStateWriter(metaStateService, startManifest);
+            globalStateGeneration = writeGlobalState(writer, newState.metaData());
+            MetaData previousMetaData = incrementalWrite ? previousClusterState.metaData() : null;
+            actions = resolveIndexMetaDataActions(previousManifest.getIndexGenerations(), getRelevantIndicesOnDataOnlyNode(newState),
+                previousMetaData, newState.metaData());
+        }
+
+        final Map<Index, Long> indexGenerations = new HashMap<>();
+        for (IndexMetaDataAction action : actions) {
+            synchronized (this) {
+                if (indicesToRetainOnDataNode.contains(action.getIndex())) {
+                    long generation = action.execute(writer);
```

Comment:
I think this could be writing some index metadata concurrently with `IndicesStore` deleting the containing folder if, for instance, a shard were unassigned from a node and then assigned back to it again: `IndicesStore` works on the (older) applied state in which the shard is not assigned whereas `indicesToRetainOnDataNode` corresponds with the (newer) committed state in which the shard is assigned again.